ext/drpttracer/drpttracer.c

```diff
@@ -0,0 +1,588 @@
+        ioctl(perf_event_fd, PERF_EVENT_IOC_ENABLE, 0) < 0) {
+        ERRMSG("Error Message: %s\n", strerror(errno));
+        ASSERT(false, "failed to start tracing");
+        pttracer_handle_free(drcontext, handle);
+        return DRPTTRACER_ERROR_FAILED_TO_START_TRACING;
+    }
+
+    *(pttracer_handle_t **)tracer_handle = handle;
+    return DRPTTRACER_SUCCESS;
+}
+
+DR_EXPORT
+drpttracer_status_t
+drpttracer_end_tracing(IN void *drcontext, INOUT void *tracer_handle,
+                       OUT drpttracer_output_t **output)
+{
+    pttracer_handle_t *handle = (pttracer_handle_t *)tracer_handle;
+    if (handle == NULL || handle->fd < 0) {
+        ASSERT(false, "invalid pttracer handle");
+        return DRPTTRACER_ERROR_INVALID_TRACING_HANDLE;
+    }
+
+    if (output == NULL) {
+        ASSERT(false, "invalid output");
+        return DRPTTRACER_ERROR_INVALID_OUTPUT;
+    }
+
+    /* Stop the pt tracing. */
+    errno = 0;
+    if (ioctl(handle->fd, PERF_EVENT_IOC_DISABLE, 0) < 0) {
+        ERRMSG("error Message: %s\n", strerror(errno));
+        ASSERT(false, "failed to stop tracing");
+        return DRPTTRACER_ERROR_FAILED_TO_STOP_TRACING;
+    }
+
+    *output =
+        (drpttracer_output_t *)dr_thread_alloc(drcontext, sizeof(drpttracer_output_t));
+    memset(*output, 0, sizeof(drpttracer_output_t));
+
+    memcpy(&(*output)->metadata, &pt_shared_metadata, sizeof(pt_metadata_t));
+    (*output)->metadata.time_shift = handle->header->time_shift;
+    (*output)->metadata.time_mult = handle->header->time_mult;
+    (*output)->metadata.time_zero = handle->header->time_zero;
+
+    if (handle->header->aux_head > handle->header->aux_size) {
+        ERRMSG("the buffer is full and the new PT data is overwritten\n");
+        ASSERT(false, "aux_head > aux_size");
+        return DRPTTRACER_ERROR_FAILED_TO_READ_TRACE;
+    }
+
+    (*output)->pt_buf_size = handle->header->aux_head - handle->header->aux_tail;
+    (*output)->pt_buf = dr_thread_alloc(drcontext, (*output)->pt_buf_size);
+    memcpy((*output)->pt_buf, (uint8_t *)handle->aux, (*output)->pt_buf_size);
+    if (handle->tracing_mode == DRPTTRACER_TRACING_ONLY_USER ||
+        handle->tracing_mode == DRPTTRACER_TRACING_USER_AND_KERNEL) {
+        if (handle->header->data_head > handle->header->data_size) {
+            ERRMSG("the buffer is full and the new PT's sideband data is overwritten\n");
+            ASSERT(false, "data_head > data_size");
+            return DRPTTRACER_ERROR_FAILED_TO_READ_SIDEBAND_DATA;
+        }
+        (*output)->sideband_buf_size =
+            handle->header->data_head - handle->header->data_tail;
+        (*output)->sideband_buf =
+            dr_thread_alloc(drcontext, (*output)->sideband_buf_size);
+        memcpy((*output)->sideband_buf,
+               (uint8_t *)handle->base + handle->header->data_tail,
+               (*output)->sideband_buf_size);
+    } else {
+        (*output)->sideband_buf = NULL;
```

Comment:
For the kernel tracing mode, can we somehow assert that there was no sideband data generated?

Reply:
In my test, the sideband buffer will record some data even if I don't set the config to record any other perf event.

Reply:
Add a comment to say that even though there is some sideband data, we don't really need it, and we're discarding it.

Reply:
Done.